### PR TITLE
feat: add inlineHtml node

### PR DESCRIPTION
**Pull request by P0lip**
Needed for https://github.com/stoplightio/markdown-viewer/pull/52

One can try it out live here https://github.com/stoplightio/markdown-viewer/pull/52

### CODE REVIEW

**Comment by P0lip** [comment]
Bah, seems like there is an empty space left in `tagName` at times.
I'm going to look into this
image

**Comment by P0lip** [comment]
urgh, I'll need to change the approach a bit and incorporate some real html-parser :disappointed: 
image
I'm marking this as draft, will ask for another pass once ready.

**Comment by marbemac** [comment]
Hmm, anything in the remark ecosystem to help us with this? Only thing I'm a little wary of is bundle size - some of the html parsers are really heavy. And if this turns out to be a huge task (like we need to write our own html parser or keep building on the stuff on this PR, maybe we revisit wether now is the time?).

File: src/plugins/__tests__/html.spec.ts [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+                    type: 'strong',
+                    position: expect.any(Object),
+                    children: [
+                      {
+                        type: 'text',
+                        value: 'hello',
+                        position: expect.any(Object),
+                      },
+                    ],
+                  },
```
Comment (marcelltoth) [inline src/plugins/__tests__/html.spec.ts]: Are you sure we want to parse MD inside HTML? 

The [SMD docs](https://meta.stoplight.io/docs/studio-demo/docs/markdown/basic-syntax.md#inline-html) say it's not supported in HTML blocks (which the docs call "inline HTML"). I'd say it shouldn't be supported here either. Consistency matters with both our block-HTML parser and [Github's parser](https://github.com/stoplightio/studio-demo/blob/master/docs/markdown/basic-syntax.md#inline-html).

File: src/ast-types/mdast.ts [hunk 2]
```diff
@@ -37,6 +38,12 @@ export interface IHTML extends Unist.Literal {
   type: 'html';
 }
 
+export interface IInlineHTML extends Unist.Parent {
+  type: 'inlineHtml';
+  attributes: Dictionary<string | true, string>;
+  tagName: string;
+}
```
Comment (marcelltoth) [inline src/ast-types/mdast.ts:45]: For the reason above, I'd say this should be treated as a literal, like `IHTML`

File: src/plugins/__tests__/html.spec.ts [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+  it('should prepare children for inline html tag', () => {
+    const parsed = parse(`HTML link <a href="https://google.com">**hello** _world_ ~~navigate~~ to</a>.`);
+
+    expect(
+      unified()
+        .use(mergeHtml)
+        .runSync(parsed),
+    ).toStrictEqual(
+      expect.objectContaining({
```
Comment (marcelltoth) [inline src/plugins/__tests__/html.spec.ts]: Don't these two contradict eachother?

If you want strict checking (which is great), why don't you simply pass an object literal?

This applies to a few other places as well.

**Review by marcelltoth — CHANGES_REQUESTED** [review]
Let's make it consistent with the other implementations.

Also, for the future if you bundle such huge (in terms of lines) refactoring with something could you do it in a separate commit please? It's a lot easier to review stuff that way, rather than being thrown at with a single +900-600 line commit.

**Comment by P0lip** [comment]
> Also, for the future if you bundle such huge (in terms of lines) refactoring with something could you do it in a separate commit please? It's a lot easier to review stuff that way, rather than being thrown at with a single +900-600 line commit.

The change is rather small - it's mostly tests that affect the amount of LOC

File: src/plugins/__tests__/html.spec.ts (hunk 1 above)
Comment (P0lip) [reply src/plugins/__tests__/html.spec.ts]: > Are you sure we want to parse MD inside HTML?

I don't know. It'd be by far easier not to, but either way, we don't need to be strict here - the code will degrade gracefully anyway, so the core SMD principle is met, but we can punt on this and tell users to use HTML tags.

The thing is that, AFAIK inline html elements can be placed among other nodes, while block html elements not.
`dl` is considered as block element.

> The SMD docs say it's not supported in HTML blocks (which the docs call "inline HTML"). I'd say it shouldn't be supported here either. Consistency matters with both our block-HTML parser and Github's parser.

`dl` are block tags/elements, unlike `span` or `a`, so the case is quite different.

Comment (P0lip) [reply src/plugins/__tests__/html.spec.ts]: Also, see this https://github.github.com/gfm/#example-138 vs https://github.github.com/gfm/#example-129

**Comment by P0lip** [comment]
I gotta admit that I could have avoided moving the jira block plugin tests :laughing:

File: src/plugins/__tests__/html.spec.ts (hunk 1 above)
Comment (P0lip) [reply src/plugins/__tests__/html.spec.ts]: Having said that, I'm pretty sure we do need to support markdown inside _inline_ html elements.
markdown under `dl` (block element) won't be treated as plain text (same as we do right now), but
`<a href="foo.com">*foo*</a>`
<a href="foo.com">*foo*</a>
will not

I won't support markdown under block elements, however, to be consistent.

File: src/plugins/__tests__/html.spec.ts (hunk 3 above)
Comment (P0lip) [reply src/plugins/__tests__/html.spec.ts]: > Don't these two contradict eachother?

Yup, true, it has no effect when paired with `objectContaining`.

> why don't you simply pass an object literal?

I can do it, I already have all properties listed anyway, so should be fairly easy to do.

File: src/ast-types/mdast.ts (hunk 2 above)
Comment (P0lip) [reply src/ast-types/mdast.ts:45]: Explained above.
There are two kind of html elements:
- inline
- block

Inline such as `span` or `a` can be placed among regular markdown markup, and markdown markup should be intereprted in such case. As demonstrated in the other coment, GH handles them as well.
Luckily, block elements do not need to be treated in the same way, so we do not need any parser nor anything.
I adjusted the code to account for that.

File: src/plugins/__tests__/html.spec.ts (hunk 1 above)
Comment (marcelltoth) [reply src/plugins/__tests__/html.spec.ts]: Alright, thanks for the explanation.

**Comment by P0lip** [comment]
so, who's willing to review again / approve? :grin:

**Review by marcelltoth — APPROVED** [review]
Eh, I'll be brave. 💪 😎

**Pull request merged**